packages/client/src/links/wsLink/wsClient/index.ts

```diff
@@ -0,0 +1,396 @@
+      ) {
+        this.inactivityTimeout.reset();
+        return;
+      }
+
+      void this.close();
+    }, lazyOptions.closeMs);
+
+    // Initialize the WebSocket connection.
+    this.activeConnection = new WsConnection({
+      WebSocketPonyfill: opts.WebSocket,
+      promiseUrl: prepareUrl(opts.url, !!opts.connectionParams).catch(
+        (error) => {
+          throw new TRPCWebSocketReconnectFatal({
+            message:
+              'Error when building url. Ensure provided url(): Promise<string> does not throws.',
+            cause: error,
+          });
+        },
+      ),
+      keepAlive: {
+        ...keepAliveDefaults,
+        ...opts.keepAlive,
+      },
+    });
+    this.activeConnection.wsObservable.subscribe({
+      next: (ws) => {
+        if (!ws) return;
+        this.setupWebSocketListeners(ws);
+      },
+    });
+
+    // Initialize the reconnection manager.
+    this.reconnectManager = new ReconnectManager(
+      opts.retryDelayMs ?? exponentialBackoff,
+      {
+        onError: (error) => {
+          this.connectionState.next({
+            type: 'state',
+            state: 'connecting',
+            error: TRPCClientError.from(error),
+          });
+        },
+      },
+    );
+    if (!lazyOptions.enabled) {
+      this.reconnectManager.attach(this.activeConnection);
+    }
+
+    this.connectionState = behaviorSubject<
+      TRPCConnectionState<TRPCClientError<AnyTRPCRouter>>
+    >({
+      type: 'state',
+      state: lazyOptions.enabled ? 'idle' : 'connecting',
+      error: null,
+    });
+
+    // Automatically open the connection if lazy mode is disabled.
+    if (!lazyOptions.enabled) {
+      void this.open();
+    }
+  }
+
+  /**
+   * Opens the WebSocket connection. Handles reconnection attempts and updates
+   * the connection state accordingly.
+   */
+  private async open() {
+    if (this.connectionState.get().state !== 'connecting') {
+      this.connectionState.next({
+        type: 'state',
+        state: 'connecting',
+        error: null,
+      });
+    }
+
+    try {
+      await this.activeConnection.open();
+    } catch (error) {
+      if (error instanceof TRPCWebSocketReconnectFatal) throw error;
+
+      this.connectionState.next({
+        type: 'state',
+        state: 'connecting',
+        error: TRPCClientError.from(
+          new TRPCWebSocketClosedError({
+            message: 'Initialization error',
+            cause: error,
+          }),
+        ),
+      });
+
+      await this.reconnectManager.reconnect();
+    }
+  }
+
+  /**
+   * Closes the WebSocket connection and stops managing requests.
+   * Ensures all pending and active requests are properly finalized.
+   */
+  public async close() {
+    this.inactivityTimeout.stop();
+    this.reconnectManager.stop();
+
+    const requestsToAwait: Promise<void>[] = [];
+    for (const {
+      state,
+      message,
+      end,
+      callbacks,
+    } of this.requestManager.getRequests()) {
```

Comment:
Remove the destructuring here, it makes it hard to read - we try to not use destructuring much unless it's variable that are referenced many times (rule of 3)

Reply:
Fixed